portal-kernel/src/com/liferay/portal/kernel/dao/model/dsl/DSLStatementUtil.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.kernel.dao.model.dsl;
+
+import com.liferay.portal.kernel.dao.model.Table;
+import com.liferay.portal.kernel.dao.model.dsl.ast.impl.DefaultASTNodeVisitor;
+import com.liferay.portal.kernel.dao.model.dsl.expressions.Expression;
+import com.liferay.portal.kernel.dao.model.dsl.query.AggregateExpression;
+import com.liferay.portal.kernel.dao.model.dsl.query.Query;
+import com.liferay.portal.kernel.dao.model.dsl.query.Select;
+import com.liferay.portal.kernel.dao.orm.QueryPos;
+import com.liferay.portal.kernel.dao.orm.SQLQuery;
+import com.liferay.portal.kernel.dao.orm.Session;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * @author Preston Crary
+ */
+public class DSLStatementUtil {
+
+	public static Select count() {
+		return new Select(
+			false,
+			new Expression[] {AggregateExpression.COUNT_STAR_COUNT_VALUE});
+	}
+
+	public static Select countDistinct(Expression<?> expression) {
+		return new Select(
+			false,
+			new Expression[] {
+				new AggregateExpression<>(true, expression, "count")
+			});
+	}
+
+	public static SQLQuery createSynchronizedSQLQuery(
+		Session session, Query query) {
+
+		DefaultASTNodeVisitor defaultASTNodeVisitor =
+			new DefaultASTNodeVisitor();
+
+		query.accept(defaultASTNodeVisitor);
+
+		Set<Table<?>> tables = defaultASTNodeVisitor.getTables();
+
+		String[] tableNames = new String[tables.size()];
+
+		int i = 0;
+
+		for (Table<?> table : tables) {
+			tableNames[i++] = table.getTableName();
+		}
+
+		SQLQuery q = session.createSynchronizedSQLQuery(
+			defaultASTNodeVisitor.toString(), true, tableNames);
+
+		List<Object> scalarValues = defaultASTNodeVisitor.getScalarValues();
+
+		if (!scalarValues.isEmpty()) {
+			QueryPos queryPos = QueryPos.getInstance(q);
+
+			for (Object value : scalarValues) {
+				queryPos.add(value);
```

Comment:
@topolik this is where user provided strings are consumed, they are set as params when creating the Hibernate query.  

Only way to manually add SQL is to define a custom keyword/node.

Reply:
Oh I see, for alias that's a whole other concern.

Reply:
Would it be acceptable to let them define the alias but verify the content of the string so if it contains anything but [A-Za-z0-9]+ we throw an exception?

Reply:
I'm afraid it's not only the alias, but all places in `DefaultASTNodeVisitor` where you append a non-constant string input into `_sb`, the table/column alias/name is just an example. Let's take the first example which would be `AggregateExpression` https://github.com/shuyangzhou/liferay-portal/pull/8675/commits/6cb80336259dacbb9c17a21ee42bf5b71c754041#diff-a8ffbc2b6d934d4171803ec1ac76da8cR92, using the public constructor devs might make a mistake and create an instance with a malicious user supplied `name` https://github.com/shuyangzhou/liferay-portal/commit/6cb80336259dacbb9c17a21ee42bf5b71c754041#diff-83cd33d048ad5c5af93b444b60d1fb38R39 and this gets directly into the SQL query.

A package protected constructor, singletons or some validating factory could solve it. In LDAP I solved it with protected constructor (https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-security/portal-security-ldap-api/src/main/java/com/liferay/portal/security/ldap/SafeLdapFilter.java#L128) and  `SafeLdapFilterFactory#fromUnsafeFilter` which accepts string but it's more obvious for devs to think twice & can be easily caught by security tooling: https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-security/portal-security-ldap-api/src/main/java/com/liferay/portal/security/ldap/SafeLdapFilterFactory.java

Reply:
Originally I was using named enums for anything like this (function types, joins, operands and aggregates).  The problem is it really closes the door for developers to define anything we do not define for them.  So if they need to use sum/max/min or another aggregate we have to define the aggregate even though we do not use it.

This is really a question of how restrictive we want to be, allowing any customization at all or forbidding customization.  From a security perspective I agree (more safe), from a developer friendly perspective I disagree (more flexible).

Reason to be more flexible is nothing is forcing anyone to use this API, any dev can grab a connection and allow SQL injection bugs.  By not being flexible enough it's more likely devs will handcode the entire SQL instead of using this API.

Reply:
Hmm, right. Can it be API vs SPI? 

In API they would not be allowed to create a new instances. If devs want to add new predicate/whatever they would use SPI. 

/ccing @csierra that wanted to be aware.